QMR Solver

Programmer-Magnus opened the pull request:
Implement QMR from SciPy, the Quasi-Minimal Residual method for matrix equations. 
It seems to work for problems with a high condition number.
Add a few tests for this new solver.

The algorithm could break down in several ways. How should this be indicated in the returned state?
Is it possible to configure preconditioning matrices? 
They are named M1 and M2 in the comments.
Should the algorithm be made to work for complex numbers?

More information about many iterative methods are found here:
https://www.netlib.org/templates/templates.pdf

Programmer-Magnus commented on src/main/java/org/ojalgo/matrix/task/iterative/QMRSolver.java:
```diff
@@ -0,0 +1,267 @@
+        // r = b - A*x
+        R064Store Ax = R064Store.FACTORY.make(n, 1L);
+        A.multiply(x).supplyTo(Ax);
+        double normRHS = ONE;
+        double normErr = ZERO;
+        for (int i = 0; i < n; i++) {
+            final double bi = b.doubleValue(i);
+            normRHS = HYPOT.invoke(normRHS, bi);
+            final double ri = bi - Ax.doubleValue(i);
+            r.set(i, ri);
+            normErr = HYPOT.invoke(normErr, ri);
+        }
```
I imitated this code from a similar one in the ConjugateGradientSolver. Is this wrong in the ConjugateGradientSolver?
I will start with ZERO instead to resolve this.

apete replied:
Sure looks wrong, and most likely I did that.

I'm beginning to think it was intentional to, as you you mentioned, avoid problems with very small `normRHS`. If you change this you have to handle (near) zero rhs some other way.

Programmer-Magnus replied:
Limited the normRHS explicitly and added a comment.

Programmer-Magnus commented on src/main/java/org/ojalgo/matrix/task/iterative/QMRSolver.java:
```diff
@@ -0,0 +1,267 @@
+import org.ojalgo.structure.Access2D;
+import org.ojalgo.type.context.NumberContext;
+
+/**
+ * Quasi-Minimal Residual (QMR) solver for general nonsymmetric square systems.
+ *
+ * This is a Java port of SciPy's qmr() reference implementation.
+ * It is almost a straight line by line translation from SciPy.
```
The GitHub perma link to the translated SciPy code is found under `References:`.

apete commented on src/main/java/org/ojalgo/matrix/decomposition/SingularValue.java:
```diff
@@ -170,7 +239,9 @@ default void ftran(final PhysicalStore<N> arg) {
 
     /**
      * @return The diagonal matrix of singular values.
+     * @deprecated Use {@link #getS()} instead
```
This change is not actually part of the PR. Do a rebase.

And I'm pretty sure that's not correct use of the @since tag/annotation.

apete replied:
You should rebase on the `develop` branch, and then the `Spectral decomposition` commit will come before your changes. You have somehow managed to put them after your own changes.

Programmer-Magnus commented:
I did not consider using it in the `ConvexSolver`, but I would be happy to modify code to make it possible. Thanks for describing how to achieve this.

If other `IterativeSolverTask` are for `double` I leave it with doubles.

I just made all indices to `long` but if you mean you would prefer `int` I will change that back, no problem. :) 

Would it be better to rebase this patch on top of 56.0.0, or something else than the present develop like I did now?

apete commented:
I'm good to merge this.

Can you rebase and squash to 1 commit before we merge?

`scaleCopy` can be implemented this way:

```
    private static void scaleCopy(final R064Store src, final double alpha, final R064Store dst) {
        OperationBinary.invoke(dst.data, 0, Math.min(src.data.length, dst.data.length), 1, alpha, MULTIPLY, src.data);
    }
```
or, like this:
```
    private static void scaleCopy(final R064Store src, final double alpha, final R064Store dst) {
        CorePrimitiveOperation.multiply(dst.data, 0, Math.min(src.data.length, dst.data.length), 1, alpha, src.data);
    }
```

Not saying it's better, nor that you should change anything. Just to show you.

Programmer-Magnus commented on src/main/java/org/ojalgo/matrix/task/iterative/QMRSolver.java:
```diff
@@ -0,0 +1,267 @@
+
+    private static void scaleCopy(final R064Store src, final double alpha, final R064Store dst) {
+        for (int i = 0; i < src.getRowDim(); i++) {
+            dst.set(i, alpha * src.doubleValue(i));
```
Ok, I removed it again.

Programmer-Magnus commented:
Commits are now rebased on top of upstream/develop and squashed to one commit.

For now I kept scaleCopy as it is because I find it easier to understand. Thanks for showing how it can be done. 

I keep the present stopping criteria (normRHS) after reading about it at netlib.org.
https://www.netlib.org/linalg/html_templates/node83.html#SECTION00920000000000000000

In the future when there is interest there will be a limited effort to make use of left and right preconditioner matrices and/or handle complex numbers instead of doubles.
